Automattic/wp-calypso
# Nav unification - auto-collapse all sections on new menu item click

This PR is created [on the back of discussions in the Issue](https://github.com/Automattic/wp-calypso/issues/46089#issuecomment-709089999) regarding which UI interaction for the Calypso-based navigation best matches that which we have in WPAdmin. 

_Please_ see that discussion _before_ providing feedback here otherwise you will be missing key context.

#### Changes proposed in this Pull Request

* Make all open navigation sections close automatically when a new section is opened.
* Allow only a single section to be open at any one time.
* Don't allow for "selected" state to determine toggle state of menu - rely on triggering the toggle based on the selected state (see the existing `useRef` implementation).
* Adds a HO reducer to the `state.mySites.sidebarSections` state slice to enable this. Should not effect existing Calypso nav as it falls through (see `default` case) to existing reducer.

#### Screenshots

Please do not rely on this screengrab for "testing" purposes. It is important the PR is tested manually in order to get a solid feel for the interaction and how it might change when clicking on nav items in different orders...etc.

![Screen Capture on 2020-10-20 at 15-34-30](https://user-images.githubusercontent.com/444434/96602366-da5b5d00-12ea-11eb-95bc-c15bfcd6ed2b.gif)


#### Testing instructions

The aim is to:

1. Validate that if a nav item (whether or not it has sub-items) is clicked any other sections that are expanded are automatically closed. It should not be possible to have x2 menu sections in an expanded state.
2. Check that the changes don't break the _existing_ Calypso nav that our users see in Production.

As hard as it is, when testing this PR, please ignore any "jumping" of the navigation causes by the "Notices/Nudges". That issue is being tackled separately.

* Checkout this PR.
* `yarn && yarn start`. Wait for Calypso to boot.
* Goto `http://calypso.localhost:3000/home/` and select your test site.
* Add `?flags=nav-unification` to the end of the URL. Hit `Enter` to reload the page.
* You should see the Unified Nav (looks like the WPAdmin nav).
* Click on `Pages` - you should see the `Pages` section expanded.
* Click on `Posts` - you should see `Pages` is auto-collapsed and `Posts` becomes expanded.
* Click on `Statistics` - you should see `Posts` auto collapse and `Statistics` is highlighted.

Also validate the UX (but in the context of the fact that Calypso is a single page application which doesn't have a full page reload to mask certain UX issues that are present in WPAdmin. Therefore it can never be 1:1 with WPAdmin):

* How closely does it mirror the WPAdmin experience?
* Is the experience intuitive?
* Are there any bugs?
* What ways could the solution be improved?
* Is this "good enough" to be merged and then iterated upon?

Fixes https://github.com/Automattic/wp-calypso/issues/46089
